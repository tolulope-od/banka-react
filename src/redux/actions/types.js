export const GET_ERRORS = 'GETTING ERRORS';
export const CLEAR_ERRORS = 'CLEARING ERRORS';
export const SET_USER = 'SETTING CURRENT USER';
