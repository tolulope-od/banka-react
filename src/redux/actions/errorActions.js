import { GET_ERRORS, CLEAR_ERRORS } from './types';

export const getErrors = error => ({ type: GET_ERRORS, payload: error });
export const clearErrors = () => ({ type: CLEAR_ERRORS });
