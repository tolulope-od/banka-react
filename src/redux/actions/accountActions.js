import axios from 'axios';
import { ACCOUNT_ACTION_LOADING, ADD_BANK_ACCOUNT, FETCH_ALL_BANK_ACCOUNTS } from './types';
import { clearErrors, getErrors } from './errorActions';
import API_URL from '../../utils/API';

const createBankAccount = type => {
  return async dispatch => {
    dispatch({ type: ACCOUNT_ACTION_LOADING, payload: true });
    try {
      dispatch(clearErrors());
      const createdAccount = await axios.post(`${API_URL}/accounts`, { type });
      const { newAccount } = createdAccount.data.data[0];
      return dispatch({ type: ADD_BANK_ACCOUNT, payload: newAccount });
    } catch (err) {
      dispatch({ type: ACCOUNT_ACTION_LOADING, payload: false });
      return dispatch(getErrors(err.response.data));
    }
  };
};

export const fetchBankAccounts = () => {
  return async dispatch => {
    dispatch({ type: ACCOUNT_ACTION_LOADING, payload: true });
    try {
      dispatch(clearErrors());
      const allAccounts = await axios.get(`${API_URL}/accounts`);
      const { data } = allAccounts.data;
      dispatch({ type: FETCH_ALL_BANK_ACCOUNTS, payload: data });
      return data;
    } catch (err) {
      dispatch({ type: ACCOUNT_ACTION_LOADING, payload: false });
      return dispatch(getErrors(err.response.data));
    }
  };
};

export default createBankAccount;
