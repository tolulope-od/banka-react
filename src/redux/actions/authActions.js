import axios from 'axios';
import jwtDecode from 'jwt-decode';
import setAuthToken from '../../utils/setAuthToken';
import { SET_USER, USER_LOADING } from './types';
import { clearErrors, getErrors } from './errorActions';
import API_URL from '../../utils/API';

export const setUser = decoded => ({ type: SET_USER, payload: decoded });

export const login = userData => {
  return async dispatch => {
    dispatch({ type: USER_LOADING, payload: true });
    try {
      dispatch(clearErrors());
      const returningUser = await axios.post(`${API_URL}/auth/signin`, userData);
      const { token } = returningUser.data.data[0];
      setAuthToken(token);
      const decoded = jwtDecode(token);
      dispatch(setUser(decoded));
      return dispatch({ type: USER_LOADING, payload: false });
    } catch (err) {
      dispatch({ type: USER_LOADING, payload: false });
      return dispatch(getErrors(err.response.data));
    }
  };
};

export const signUp = (userData, history) => {
  return async dispatch => {
    dispatch({ type: USER_LOADING, payload: true });
    try {
      dispatch(clearErrors());
      const newUser = await axios.post(`${API_URL}/auth/signup`, userData);
      const { token } = newUser.data.data[0];
      setAuthToken(token);
      const decoded = jwtDecode(token);
      dispatch(setUser(decoded));
      dispatch({ type: USER_LOADING, payload: false });
      return history.push('/createAccount');
    } catch (err) {
      dispatch({ type: USER_LOADING, payload: false });
      return dispatch(getErrors(err.response.data));
    }
  };
};
