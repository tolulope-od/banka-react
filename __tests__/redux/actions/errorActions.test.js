import configureMockStore from 'redux-mock-store';
import { getErrors, clearErrors } from '../../../src/redux/actions/errorActions';

const mockStore = configureMockStore();
const store = mockStore({});

describe('Error action creators', () => {
  it('creates a GET_ERRORS action whenever an error occurs', () => {
    store.dispatch(
      getErrors({
        status: 404,
        error: 'Account not found!'
      })
    );
    expect(store.getActions()).toMatchSnapshot();
  });

  it('creates a CLEAR_ERRORS action to clear the errors', () => {
    store.dispatch(clearErrors());
    expect(store.getActions()).toMatchSnapshot();
  });
});
