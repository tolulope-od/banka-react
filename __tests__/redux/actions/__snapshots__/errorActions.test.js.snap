// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error action creators creates a CLEAR_ERRORS action to clear the errors 1`] = `
Array [
  Object {
    "payload": Object {
      "error": "Account not found!",
      "status": 404,
    },
    "type": "GETTING ERRORS",
  },
  Object {
    "type": "CLEARING ERRORS",
  },
]
`;

exports[`Error action creators creates a GET_ERRORS action whenever an error occurs 1`] = `
Array [
  Object {
    "payload": Object {
      "error": "Account not found!",
      "status": 404,
    },
    "type": "GETTING ERRORS",
  },
]
`;
